source/fundamentals/connection.txt

```diff
@@ -0,0 +1,434 @@
+       thread may wait for a connection to become available.
+
+   * - **serverSelectionTimeoutMS**
+     - integer
+     - Specifies the maximum amount of time, in milliseconds, the driver
+       will wait for server selection to succeed before throwing an
+       exception.
+
+   * - **localThresholdMS**
+     - integer
+     - When communicating with multiple instances of MongoDB in a replica
+       set, the driver will only send requests to a server whose
+       response time is less than or equal to the server with the fastest
+       response time plus the local threshold, in milliseconds.
+
+   * - **heartbeatFrequencyMS**
+     - integer
+     - Specifies the frequency, in milliseconds, that the driver will
+       wait between attempts to determine the current state of each
+       server in the cluster.
+
+   * - **replicaSet**
+     - string
+     - Specifies that the :ref:`connection string <connection-uri>`
+       provided includes multiple hosts. When specified, the driver
+       attempts to find all members of that set.
+
+   * - **ssl**
+     - boolean
+     - Specifies that all communication with MongoDB instances should
+       use TLS/SSL. Superseded by the **tls** option.
+
+   * - **tls**
+     - boolean
+     - Specifies that all communication with MongoDB instances should
+       use TLS. Supersedes the **ssl** option.
+
+   * - **tlsInsecure**
+     - boolean
+     - Specifies that the driver should allow invalid hostnames for TLS
+       connections. Has the same effect as setting
+       **tlsAllowInvalidHostnames** to true. To configure TLS security
+       constraints in other ways, use a
+       :ref:`custom SSLContext <tls-custom-sslContext>`.
+
+   * - **sslInvalidHostNameAllowed**
+     - boolean
+     - Specifies that the driver should allow invalid hostnames in the
+       certificate for TLS/SSL connections. Superseded by
+       **tlsAllowInvalidHostnames**.
+
+   * - **tlsAllowInvalidHostnames**
+     - boolean
+     - Specifies that the driver should allow invalid hostnames in the
+       certificate for TLS connections. Supersedes
+       **sslInvalidHostNameAllowed**.
+
+   * - **connectTimeoutMS**
+     - integer
+     - Specifies the maximum amount of time, in milliseconds, the Java
+       driver will wait for a connection to open before timing out.
+
+   * - **socketTimeoutMS**
+     - integer
+     - Specifies the maximum amount of time, in milliseconds, the Java
+       driver will wait to send or receive a request before timing out.
+
+   * - **maxIdleTimeMS**
+     - integer
+     - Specifies the maximum amount of time, in milliseconds, the Java
+       driver will allow a pooled connection to idle before closing the
+       connection.
+
+   * - **maxLifeTimeMS**
+     - integer
+     - Specifies the maximum amount of time, in milliseconds, the Java
+       driver will continue to use a pooled connection before closing the
+       connection.
+
+   * - **safe**
```

Comment:
I'm not sure we need to document this,  writeconcern is a better parameter for safety.

Reply:
Agreed. I'd like to remove this, but it is mentioned in quite a few other option descriptions. I worry that those descriptions would feel incomplete without that mention (or without a definition for the **safe** option). The same thing happened with the **ssl** option -- do you think we should hide both of these entirely, or do you think it's OK to keep them around since they have historical value?

Reply:
Regarding ssl - I think that might be useful as not all users know the difference between ssl and tls.

I'm not sure where else safe is used, so happy to leave the decision to you. To note: there is a link at the end to all options via the java doc.

Reply:
I also think that `safe` should be removed.  Especially as it just means `w=1`, which is the default anyway.

Reply:
It also implies something that is not true: w: 1 writes can be lost on a rollback